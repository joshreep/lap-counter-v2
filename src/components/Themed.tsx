/**
 * Learn more about Light and Dark modes:
 * https://docs.expo.io/guides/color-schemes/
 */

import { HTMLAttributes } from 'react'
import styled from 'styled-components'

type ThemeProps = {
  lightColor?: string
  darkColor?: string
}

export type ViewProps = HTMLAttributes<HTMLDivElement> & ThemeProps & { $rootBackground?: boolean }

export const TextInput = styled.input`
  color: ${({ theme }) => theme.colors.text};
  background-color: ${({ theme }) => theme.colors.inputBackground};
  border-radius: 4px;
`

type TextTags = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'p' | 'span'

export const Text = (tag: TextTags = 'p') => styled(tag)`
  color: ${({ theme }) => theme.colors.text};
`

export const View = styled.div<ViewProps>`
  background-color: ${(props) =>
    ((props.theme.dark ? props.darkColor : props.lightColor) ?? props.$rootBackground)
      ? props.theme.colors.rootBackground
      : props.theme.colors.background};
`
