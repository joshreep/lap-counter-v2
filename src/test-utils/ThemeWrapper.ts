import MatchMediaMock from 'jest-matchmedia-mock'

type Theme = 'light' | 'dark'

const themes: Theme[] = ['light', 'dark']

export function forEachTheme(callback: (theme: Theme) => void) {
  themes.forEach((theme) => {
    const matchMedia = new MatchMediaMock()
    matchMedia.useMediaQuery(`(prefers-color-scheme: ${theme})`)

    callback(theme)

    matchMedia.clear()
  })
}
