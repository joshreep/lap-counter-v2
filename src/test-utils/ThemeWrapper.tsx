import SCDarkTheme from '@/theme/dark-theme'
import SCLightTheme from '@/theme/light-theme'
import {
  DarkTheme as RNDarkTheme,
  DefaultTheme as RNDefaultTheme,
  ThemeProvider as RNThemeProvider,
} from '@react-navigation/native'
import { FC, PropsWithChildren } from 'react'
import { ColorSchemeName } from 'react-native'
import { ThemeProvider as SCThemeProvider } from 'styled-components/native'

export const themes: ColorSchemeName[] = ['light', 'dark']

export function getThemeWrapper(theme: ColorSchemeName) {
  const Wrapper: FC<PropsWithChildren> = ({ children }) => {
    return (
      <RNThemeProvider value={theme === 'dark' ? RNDarkTheme : RNDefaultTheme}>
        <SCThemeProvider theme={theme === 'dark' ? SCDarkTheme : SCLightTheme}>{children}</SCThemeProvider>
      </RNThemeProvider>
    )
  }

  return Wrapper
}

type ForEachThemeCallback = (theme: ColorSchemeName, wrapper: FC<PropsWithChildren>) => void

export function forEachTheme(callback: ForEachThemeCallback) {
  themes.forEach((theme) => {
    callback(theme, getThemeWrapper(theme))
  })
}
