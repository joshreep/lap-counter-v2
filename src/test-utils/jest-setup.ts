import '@testing-library/react-native/extend-expect'
import '@testing-library/jest-native/extend-expect'

jest.mock('react-native-reanimated', () => null, { virtual: true })
jest.mock('@testing-library/jest-native/extend-expect', () => null, { virtual: true })
