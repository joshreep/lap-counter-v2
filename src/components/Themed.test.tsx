import { forEachTheme } from '@/test-utils/ThemeWrapper'
import { render } from '@testing-library/react-native'
import { Text, TextInput, View } from './Themed'

forEachTheme((theme, wrapper) => {
  describe(`${theme} theme`, () => {
    test('TextInput should match snapshot', () => {
      const container = render(<TextInput value="Check, check. Is this thing on?" />, { wrapper })
      expect(container.toJSON()).toMatchSnapshot()
    })

    test('Text should match snapshot', () => {
      const container = render(<Text>Hello there</Text>, { wrapper })
      expect(container.toJSON()).toMatchSnapshot()
    })

    test('View should match snapshot', () => {
      const container = render(
        <View>
          <Text>Hello there</Text>
        </View>,
        { wrapper },
      )
      expect(container.toJSON()).toMatchSnapshot()
    })
  })
})

describe('TextInput', () => {})
