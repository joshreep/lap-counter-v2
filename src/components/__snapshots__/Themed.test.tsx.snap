// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dark theme Text should match snapshot 1`] = `
<Text
  style={
    {
      "color": "rgb(255, 255, 255)",
    }
  }
>
  Hello there
</Text>
`;

exports[`dark theme TextInput should match snapshot 1`] = `
<TextInput
  style={
    {
      "backgroundColor": "rgba(255,255,255, .1)",
      "borderBottomLeftRadius": 4,
      "borderBottomRightRadius": 4,
      "borderTopLeftRadius": 4,
      "borderTopRightRadius": 4,
      "color": "rgb(255, 255, 255)",
    }
  }
  value="Check, check. Is this thing on?"
/>
`;

exports[`dark theme View should match snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "rgba(255, 255, 255, .07)",
    }
  }
>
  <Text
    style={
      {
        "color": "rgb(255, 255, 255)",
      }
    }
  >
    Hello there
  </Text>
</View>
`;

exports[`light theme Text should match snapshot 1`] = `
<Text
  style={
    {
      "color": "#111",
    }
  }
>
  Hello there
</Text>
`;

exports[`light theme TextInput should match snapshot 1`] = `
<TextInput
  style={
    {
      "backgroundColor": "rgba(0, 0, 0, .07)",
      "borderBottomLeftRadius": 4,
      "borderBottomRightRadius": 4,
      "borderTopLeftRadius": 4,
      "borderTopRightRadius": 4,
      "color": "#111",
    }
  }
  value="Check, check. Is this thing on?"
/>
`;

exports[`light theme View should match snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "rgba(0, 0, 0, .05)",
    }
  }
>
  <Text
    style={
      {
        "color": "#111",
      }
    }
  >
    Hello there
  </Text>
</View>
`;
